Gems/MotionMatching/Code/Source/PoseDataJointVelocities.cpp

```diff
@@ -138,15 +140,70 @@ namespace EMotionFX::MotionMatching
 
     void PoseDataJointVelocities::CalculateVelocity(MotionInstance* motionInstance, size_t relativeToJointIndex)
     {
+        const size_t numJoints = m_velocities.size();
         SetRelativeToJointIndex(relativeToJointIndex);
         ActorInstance* actorInstance = motionInstance->GetActorInstance();
-        m_velocities.resize(actorInstance->GetNumNodes());
-        m_angularVelocities.resize(actorInstance->GetNumNodes());
-        for (size_t i = 0; i < m_velocities.size(); ++i)
+        m_velocities.resize(numJoints);
+        m_angularVelocities.resize(numJoints);
+
+        const float originalTime = motionInstance->GetCurrentTime();
+
+        // Prepare for sampling.
+        AnimGraphPosePool& posePool = GetEMotionFX().GetThreadData(actorInstance->GetThreadIndex())->GetPosePool();
+        AnimGraphPose* prevPose = posePool.RequestPose(actorInstance);
+        AnimGraphPose* currentPose = posePool.RequestPose(actorInstance);
+        Pose* bindPose = actorInstance->GetTransformData()->GetBindPose();
+
+        const size_t numSamples = 3;
+        const float timeRange = 0.05f; // secs
+        const float halfTimeRange = timeRange * 0.5f;
+        const float startTime = originalTime - halfTimeRange;
+        const float frameDelta = timeRange / numSamples;
+        const float motionDuration = motionInstance->GetMotion()->GetDuration();
+
+        // Zero all linear and angular velocities.
+        Reset();
+
+        for (size_t sampleIndex = 0; sampleIndex < numSamples + 1; ++sampleIndex)
+        {
+            float sampleTime = startTime + sampleIndex * frameDelta;
+            sampleTime = AZ::GetClamp(sampleTime, 0.0f, motionDuration);
+
+            if (sampleIndex == 0)
+            {
+                motionInstance->SetCurrentTime(sampleTime);
```

Comment:
nit: you can take motionInstance->SetCurrentTime outside of the if loop.

Reply:
Good call, changed